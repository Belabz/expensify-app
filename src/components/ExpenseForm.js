import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

// Create a moment date - Actual time
const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			note: props.expense ? props.expense.note : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			formError: ''
		};
	}

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }))
	};

	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }))
		}
	};

	onNoteChange = (e) => {
		// const note = e.target.value;
		// Optional
		e.persist();
		this.setState(() => ({ note: e.target.value }))
	};

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }))
		}
	};

	onFocusedChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }))
	};

	onSubmit = (e) => {
		e.preventDefault();

		// Generate a form error
		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ formError: 'Please provide a description and amount!' }));
		} else {
			this.setState(() => ({ formError: ''}));

			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
	};

	render() {
		return (
			<div>
				{ this.state.formError && <p>{ this.state.formError }</p> }

				<form onSubmit={ this.onSubmit }>
					<input
						type="text"
						placeholder="Description"
						value={ this.state.description }
						onChange={ this.onDescriptionChange }
						autoFocus
					/>

					<input
						type="text"
						placeholder="Amount"
						value={ this.state.amount }
						onChange={ this.onAmountChange }
					/>

					<SingleDatePicker
						date={ this.state.createdAt }
						onDateChange={ this.onDateChange }
						focused={ this.state.calendarFocused }
						onFocusChange={ this.onFocusedChange }
						numberOfMonths={ 1 }
						isOutsideRange={ () => false }
					/>

					<textarea
						placeholder="Add a note for your expense (Optional)"
						value={ this.state.note }
						onChange={ this.onNoteChange }
					>
					</textarea>

					<button>Add Expense</button>
				</form>
			</div>
		)
	}
}
