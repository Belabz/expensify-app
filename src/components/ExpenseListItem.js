import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
	<div>
		<Link to={`/edit/${ id }`}>
			<h3>{ description }</h3>
		</Link>
		<p>
			{ numeral(amount / 100).format('$0,0.00') }
			-
			{ moment(createdAt).format('MMMM Do, YYYY') }
		</p>
	</div>
);

// In the first parenthesis we don't need anything from the state
// But we only provide the component ExpenseListItem.
export default ExpenseListItem;
