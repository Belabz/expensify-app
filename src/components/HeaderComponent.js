import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../redux/actions/auth'

/* -------------------------------------------------------------------------
   - adding export for testing purposes
------------------------------------------------------------------------- */
export const HeaderComponent = ({ logout }) => (
	<header>
		<h1>Expensify</h1>
		<NavLink to="/" activeClassName="is-active" exact={ true }>Dashboard</NavLink>
		<NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
		<NavLink to="/help" activeClassName="is-active">Help</NavLink>
		<button onClick={ logout }>Logout</button>
	</header>
);

/* -------------------------------------------------------------------------
   - Connection to redux, setting the two calls ()()
   - The first call is for the piece for state we need or dispatchers
     we want to use
   - For the first argument we don't need any state (undefined)
   - For the second argument we provide an map dispatch to props function
   - The second call is the component we try to connect
------------------------------------------------------------------------- */
const mapDispatchToProps = (dispatch) => ({
	logout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(HeaderComponent);
