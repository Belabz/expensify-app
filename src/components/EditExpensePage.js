import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../redux/actions/expenses';

import ExpenseForm from './ExpenseForm';

const EditExpensePage = (props) => {
	// console.log(props);
	return (
		<div>
			<ExpenseForm
				expense={ props.expense }
				onSubmit={(expense) => {
					props.dispatch(editExpense(props.expense.id, expense));
					// Redirect to home page
					props.history.push('/')
				}}
			/>

			<button onClick={() => {
				props.dispatch(removeExpense({ id: props.expense.id }));
				props.history.push('/')
			}}>Remove</button>
		</div>
	);
};

const mapStateToProps = (state, props) => {
	return {
		// Array find method, find allows us to search through
		// an array looking for a single item.
		// We determine whether or not we found the correct item
		// by returning true or false from the callback.
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	}
};

export default connect(mapStateToProps)(EditExpensePage);
