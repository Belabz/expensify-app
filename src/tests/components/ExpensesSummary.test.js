import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should correctly render expenses summary with one expense', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={235}/>);
	expect(wrapper).toMatchSnapshot()
});

test('Should correctly render expenses summary with multiple expense', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={23} expensesTotal={235123123}/>);
	expect(wrapper).toMatchSnapshot()
});
