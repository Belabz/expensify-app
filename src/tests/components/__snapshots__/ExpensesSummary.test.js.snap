// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render expenses summary with multiple expense 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $2,351,231.23
  </h1>
</div>
`;

exports[`Should correctly render expenses summary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;
